File: haskell/2025/day04a/src/Main.hs
module Main ( main ) where

import Control.DeepSeq
import Data.Ix
import qualified Data.Set as Set
import System.Clock
import System.Environment
import System.Exit
import System.IO

process :: String -> Int
process content =
    let ls = lines content
    in case ls of
        [] -> error "empty input"
        (first:_) ->
            let rows = length ls
                cols = length first
                bounds = ((0, 0), (rows - 1, cols - 1))
                rolls = Set.fromList[ (r, c) | r <- [0 .. rows - 1], c <- [0 .. cols - 1], ls !! r !! c == '@' ]
                neighbors (r, c) =
                    [ neighbor | (dr, dc) <- [ (-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1) ]
                               , let neighbor = (r + dr, c + dc)
                               , bounds `inRange` (r + dr, c + dc)
                               , neighbor `Set.member` rolls
                    ]
                accessible = Set.filter (\roll -> length (neighbors roll) < 4) rolls
            in length accessible

usage :: String -> IO ()
usage progname = do
    hPutStrLn stderr $ "usage: " ++ progname ++ " <input file>"
    exitFailure

showTime :: TimeSpec -> String
showTime elapsed =
    let ns = fromIntegral (toNanoSecs elapsed) :: Double
    in if ns < 1000
       then show ns ++ " ns"
       else if ns < 1000000
       then show (ns / 1000.0) ++ " μs"
       else if ns < 10000000000
            then show (ns / 1000000.0) ++ " ms"
            else show (ns / 1000000000.0) ++ " s"

main :: IO ()
main = do
    args <- getArgs
    progname <- getProgName
    case args of
        [filename] -> do
            start <- getTime Monotonic
            content <- readFile filename
            let result = process content
            result `deepseq` return ()
            end <- getTime Monotonic
            let elapsed = diffTimeSpec start end
            putStrLn $ "result = " ++ show result
            putStrLn $ "elapsed time: " ++ showTime elapsed
        _ -> usage progname

File: rust/2025/day04a/src/main.rs
use std::collections::HashSet;
use std::env;
use std::fs;
use std::io;
use std::process;
use std::time::Instant;

fn usage(progname: &str) {
    eprintln!("usage: {progname} <input file>");
    process::exit(1);
}

fn process_file(filename: &str) -> io::Result<i64> {
    let contents = fs::read_to_string(filename)?;
    let mut rolls = HashSet::new();
    for (row, line) in contents.lines().enumerate() {
        for (col, &byte) in line.as_bytes().iter().enumerate() {
            if byte == b'@' {
                rolls.insert((row as i64, col as i64));
            }
        }
    }
    const NEIGHBORS: [(i64, i64); 8] = [
        (-1, -1),
        (-1, 0),
        (-1, 1),
        (0, -1),
        (0, 1),
        (1, -1),
        (1, 0),
        (1, 1),
    ];
    let accessible = rolls
        .iter()
        .filter(|pos| {
            let (row, col) = **pos;
            let neighbor_count = NEIGHBORS
                .iter()
                .map(|(dr, dc)| (row + dr, col + dc))
                .filter(|neighbor| rolls.contains(neighbor))
                .count();
            neighbor_count < 4
        })
        .count();
    Ok(accessible as i64)
}

fn display_duration(duration: u128) {
    print!("elapsed time: ");
    let duration_f = duration as f64;
    if duration_f < 1_000.0 {
        println!("{} ns", duration_f);
    } else if duration_f < 1_000_000.0 {
        println!("{} μs", duration_f / 1_000.0);
    } else if duration_f < 1_000_000_000.0 {
        println!("{} ms", duration_f / 1_000_000.0);
    } else {
        println!("{} s", duration_f / 1_000_000_000.0);
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let progname = &args[0];
    if args.len() < 2 {
        usage(progname);
    }
    let filename = &args[1];
    let start = Instant::now();
    match process_file(filename) {
        Ok(result) => {
            let duration = start.elapsed().as_nanos();
            println!("result = {result}");
            display_duration(duration);
        }
        Err(e) => {
            eprintln!("error while processing file `{filename}': {e}");
            process::exit(1);
        }
    }
}
